<div style="text-align: center"> <H2> Домашнее задание №15, к лекции "Базовые понятия статистики" </H2></div>
<div style='text-align: center'> <H4> по курсу PYDA-15 - Python для анализа данных </H4></div>
</div>
<p>
<div style="text-align: right"> Выполнил: студент группы PAL-4 Константин Лемешок </div>

---

Будем осуществлять работу с набором данных о состоянии здоровья лошадей, испытывающих кишечные колики (horse_data.csv).

In [1]:
import pandas as pd

In [2]:
names_full = ['surgery', 'age', 'hospital_number', 'rectal_temperature', 'pulse', 'respiratory_rate', 
              'temperature_of_extremities', 'peripheral_pulse', 'mucous_membranes', 'capillary_refill_time', 
              'pain', 'peristalsis', 'abdominal_distension', 'nasogastric_tube', 'nasogastric_reflux', 
              'nasogastric_reflux_PH', 'rectal_examination_-_feces', 'abdomen', 'packed_cell_volume', 
              'total_protein', 'abdominocentesis_appearance', 'abdominocentesis_total_protein', 'outcome', 
              'surgical_lesion?', 'type_of_lesion_1', 'type_of_lesion_2', 'type_of_lesion_3', 'cp_data']

In [3]:
df = pd.read_csv('horse_data.csv', names=names_full, na_values='?')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   surgery                         299 non-null    float64
 1   age                             300 non-null    int64  
 2   hospital_number                 300 non-null    int64  
 3   rectal_temperature              240 non-null    float64
 4   pulse                           276 non-null    float64
 5   respiratory_rate                242 non-null    float64
 6   temperature_of_extremities      244 non-null    float64
 7   peripheral_pulse                231 non-null    float64
 8   mucous_membranes                253 non-null    float64
 9   capillary_refill_time           268 non-null    float64
 10  pain                            245 non-null    float64
 11  peristalsis                     256 non-null    float64
 12  abdominal_distension            244 

#### Задание 1. Базовое изучение
<div class="alert alert-block alert-success">

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

</div>

---

**Определим тип каждого из параметров.**</p>
И представим рассуждения, относительно целесообразности включения параметра в ограниченный перечень из 8 параметров для дальнейшего изучения.

**Критерии для выбора:**
- количество данных по этому параметру;
- субъективная значимость параметра;
- тип параметра (чтобы в отобранном датафрейме были представлены данные разных типов).

---

**0. surgery** - категориальная, номинальная величина</p>
*По этому параметру есть данные для 299 случаев из 300. Параметр субъективно значим, т.к. позволяет разделить все случаи на 2 типа: "терапию" и "хирургию". Параметр в датафрейм для дальнейшего изучения **включаем**.*</p>

**1. age** - категориальная, порядковая величина</p>
*По этому параметру есть данные для всех 300 случаев. Параметр субъективно значим, т.к. позволит проверить гипотезу о том, что исход лечения зависит от возраста лошади. Параметр в датафрейм для дальнейшего изучения **включаем**.*</p>

**2. hospital_number** - категориальная, номинальная величина</p>
*По этому параметру есть данные для всех 300 случаев. Но это всего лишь учетный номер, и непонятно, зачем он нам. Параметр в датафрейм для дальнейшего изучения **не включаем**.*</p>

**3. rectal_temperature** - количественная, непрерывная величина</p>
*По этому параметру есть данные для 240 случаев из 300. Это субъективно значимый параметр, т.к. хорошо характеризует состояние организма. Параметр в датафрейм для дальнейшего изучения **включаем**.*</p>

**4. pulse** - количественная, дискретная величина</p>
*По этому параметру есть данные для 276 случаев из 300. Параметр субъективно значим, т.к. отражает общее состояние сердечно-сосудистой системы лошади, повышенный пульс может свидетельствовать об острой недостаточности кровообращения. Параметр в датафрейм для дальнейшего изучения **включаем**.*</p>

**5. respiratory_rate** - количественная, дискретная величина</p>
*По этому параметру есть данные для 242 случаев из 300. Опираясь на жизненный опыт, можем сказать, что будет наблюдаться положительная корреляция между 'respiratory_rate' и 'pulse'. Ввиду лимита на количество столбцов, параметр в датафрейм для дальнейшего изучения **не включаем**.*</p>

**6. temperature_of_extremities** - категориальная, номинальная величина</p>
*Вероятно, можно эту величину представить как категориальную порядковую, ранжировав ее значения следующим образом: cold-cool-normal-warm. Однако в представленном к заданию описании датасета такого ранжирования не наблюдается. Поэтому считаем величину номинальной. По этому параметру есть данные для 244 случаев из 300. Субъективно, параметр может быть значим в процессе лечения. Для пост-анализа, как в нашем случае, лучше подойдут более точные количественные показатели. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**7. peripheral_pulse** - категориальная, номинальная величина</p>
*Аргументация выбора типа величины аналогична п.6. По этому параметру есть данные для 231 случаев из 300. Субъективно, параметр может быть значим в процессе лечения. Для пост-анализа, как в нашем случае, лучше подойдут более точные количественные показатели. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**8. mucous_membranes** - категориальная, номинальная величина</p>
*По этому параметру есть данные для 253 случаев из 300. Субъективно, параметр может быть значим в процессе лечения. Для пост-анализа, как в нашем случае, лучше подойдут более точные количественные показатели. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**9. capillary_refill_time** - категориальная, порядковая величина</p>
*По этому параметру есть данные для 268 случаев из 300. Субъективно, параметр может быть значим в процессе лечения. Для пост-анализа, как в нашем случае, лучше подойдут более точные количественные показатели, в частности 'pulse', который также характеризует кровообращение. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**10. pain** - категориальная, номинальная величина</p>
*По этому параметру есть данные для 245 случаев из 300. Субъективно, параметр может быть значим в процессе лечения, в частности, являться показанием для хирургического вмешательства. Для пост-анализа, как в нашем случае, лучше подойдут более объективные количественные показатели. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**11. peristalsis** - категориальная, номинальная величина</p>
*Несмотря на то, что значения величины ранжированы в зависимости от степени подвижности перистальтики (от 1 до 4 в сторону уменьшения), оптимальным является одно из средних значений. Поэтому при определении типа величины применим ту же аргументацию, что и в п.6. По этому параметру есть данные для 256 случаев из 300. Субъективно, параметр может быть значим в процессе лечения, в частности, являться показанием для хирургического вмешательства. Для пост-анализа, как в нашем случае, лучше подойдут более точные количественные показатели. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**12. abdominal_distension** - категориальная, порядковая величина</p>
*Здесь оптимальным является крайнее (первое) значение. Следующие за ним значения последовательно "ухудшаются". Таким образом мы можем говорить об их порядке. По этому параметру есть данные для 244 случаев из 300. Т.к. в описании датасета этот параметр отмечен как важный, **включаем** его в датафрейм для дальнейшего исследования.*</p>

**13. nasogastric_tube** - категориальная, порядковая величина</p>
*Аргументация для определения типа аналогична п.12. По этому параметру есть данные для 196 случаев из 300, что меньше, чем для многих других параметров. Субъективная значимость этого параметра также относительно низкая. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**14. nasogastric_reflux** - категориальная, номинальная величина</p>
*Можно было сделать ее порядковой, если значения 2 и 3 поменять местами. А так - номинальная. По этому параметру есть данные для 194 случаев из 300, что меньше, чем для многих других параметров. Субъективная значимость этого параметра также относительно низкая. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**15. nasogastric_reflux_PH** - количественная, дискретная величина</p>
*По этому параметру есть данные для 53 случаев из 300, что совсем мало, по сравнению с другими показателями. Кроме этого, его субъективная значимость также не самая высокая. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**16. rectal_examination_-_feces** - категориальная, номинальная величина</p>
*По этому параметру есть данные для 198 случаев из 300. Субъективно, параметр может быть значим в процессе лечения, в частности может свидетельствовать о непроходимости кишечника и являться показанием для хирургического вмешательства. Для пост-анализа, как в нашем случае, лучше подойдут более точные количественные показатели. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**17. abdomen** - категориальная, номинальная величина</p>
*По этому параметру есть данные для 182 случаев из 300. Субъективно, параметр может быть значим в процессе лечения, в частности может характеризовать степень проходимости кишечника и являться показанием для хирургического вмешательства. Для пост-анализа, как в нашем случае, лучше подойдут более точные количественные показатели. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**18. packed_cell_volume** - количественная, дискретная величина</p>
*Дискретная потому - что количество клеток - целое значение. По этому параметру есть данные для 271 случаев из 300. Субъективно этот параметр видится значимым. Кроме того, его тип "сбалансирует" наш датафрейм с точки зрения представленности данных разных типов. Поэтому **включаем** этот параметр в датафрейм для дальнейшего изучения. *</p>

**19. total_protein** - количественная, непрерывная величина</p>
*Непрерывная потому, что рассчитывается по формуле, которая является отношением, а отношения не имеют конечной точности измерения. По этому параметру есть данные для 267 случаев из 300. Субъективно этот параметр видится значимым. Кроме того, его тип "сбалансирует" наш датафрейм с точки зрения представленности данных разных типов. Поэтому **включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**20. abdominocentesis_appearance** - категориальная, номинальная величина</p>
*В случае, если значение 'cloudy' является промежуточным между 'clear' и 'serosanguinous', величину можно считать порядковой. Однако мы не знаем, так ли это, из-за недостаточного владения предметной областью. Поэтому принимаем решение считать величину номинальной. По этому параметру есть данные для 135 случаев из 300, что гораздо меньше, чем у большинства других параметров. Кроме этого, субъективно параметр менее значим, чем многие другие. Поэтому **не включаем** этот параметр в датафрейм для дальнейшего изучения.*</p>

**21. abdominocentesis_total_protein** - количественная, непрерывная величина</p>
*Тип аналогичен п.19. По этому параметру есть данные для 102 случаев из 300, что гораздо меньше, чем у большинства других параметров. Т.к. мы уже взяли для итогового датафрейма один "протеиновый" показатель (см. п. 19), этот **не включаем**.*</p>

**22. outcome** - категориальная, номинальная величина</p>
*По этому параметру есть данные для 299 случаев из 300. Параметр субъективно значим, т.к. было бы хорошо знать, что в итоге слуилось с лошадью. Параметр в датафрейм для дальнейшего изучения **включаем**.*</p>

**23. surgical_lesion?** - категориальная, номинальная величина</p>
*По этому параметру есть данные для всех 300 случаев. Параметр субъективно достаточно значим, однако ввиду ограниченности размеров итогового датафрейма мы его **не включаем**, отдавая предпочтение другому "хирургическому" параметру 'surgery'.*</p>

**24. type_of_lesion_1, 25. type_of_lesion_2, 27. type_of_lesion_3** - категориальная, номинальная величина</p>
*По каждому из этих параметров есть данные для всех 300 случаев. При этом для 'type_of_lesion_2' данные нулевые для 293 из 300 случаев, а для 'type_of_lesion_3' - для 299 из 300 случаев. Поэтому параметры 'type_of_lesion_2' и 'type_of_lesion_3' **не включаем** в итоговый датафрейм из-за категорически малого объема данных.*</p>

*При этом для 'type_of_lesion_1' данных относительно достаточно: ненулевыми являются 244 значений из 300. Однако и эту колонку мы **не включаем** в итоговый датафрейм из-за ее типа. Ведь нам нужно будет заполнить пропуски для 56 значений. А в случае с номинальными величинами оптимальный вариант - заполнение модой. Но т.к. пропусков в колонке около 19%, заполнение модой существенно исказит другие средние показатели. Т.е. колонка в целом "проблемная" и лучше с ней на первом этапе не связываться, раз есть такая возможность :)*

**27. cp_data** - категориальная, номинальная величина</p>
*По этому параметру есть данные для всех 300 случаев. Однако, согласно описанию к датасету, эта величина незначима. **Не включаем** ее.*</p>

Таким образом, для дальнейшего исследования формируем сбалансированный по **типам**, **количеству** и **значимости** данных датафрейм из следующих столбцов:

- age: категориальная, порядковая величина
- rectal_temperature: количественная, непрерывная величина
- pulse: количественная, дискретная величина
- packed_cell_volume: количественная, дискретная величина
- total_protein: количественная, непрерывная величина
- abdominal_distension: категориальная, порядковая величина
- surgery: категориальная, номинальная величина
- outcome: категориальная, номинальная величина

In [4]:
selected_df = df[['age', 'rectal_temperature', 'pulse', 'packed_cell_volume', 'total_protein', 'abdominal_distension', 'surgery', 'outcome' ]]
selected_df.head()

age  rectal_temperature  pulse  packed_cell_volume  total_protein  \
0    1                38.5   66.0                45.0            8.4   
1    1                39.2   88.0                50.0           85.0   
2    1                38.3   40.0                33.0            6.7   
3    9                39.1  164.0                48.0            7.2   
4    1                37.3  104.0                74.0            7.4   

   abdominal_distension  surgery  outcome  
0                   4.0      2.0      2.0  
1                   2.0      1.0      3.0  
2                   1.0      2.0      1.0  
3                   4.0      1.0      2.0  
4                   NaN      2.0      2.0

**Чтобы Pandas не считал наш датафрейм копией и в дальнейшем не выдавал [неуместных предупреждений](https://prnt.sc/weu5qr), сохраним его как csv-файл и прочитаем еще раз.** Другого способа, как убедить Pandas в "оригинальности" датафрейма, пока не знаю.

In [5]:
selected_df.to_csv('horse_data_selected.csv', index=False)

In [6]:
selected_df = pd.read_csv('horse_data_selected.csv')
selected_df.head()

age  rectal_temperature  pulse  packed_cell_volume  total_protein  \
0    1                38.5   66.0                45.0            8.4   
1    1                39.2   88.0                50.0           85.0   
2    1                38.3   40.0                33.0            6.7   
3    9                39.1  164.0                48.0            7.2   
4    1                37.3  104.0                74.0            7.4   

   abdominal_distension  surgery  outcome  
0                   4.0      2.0      2.0  
1                   2.0      1.0      3.0  
2                   1.0      2.0      1.0  
3                   4.0      1.0      2.0  
4                   NaN      2.0      2.0

Рассчитаем базовые метрики для количественных величин и кратко опишем результаты:

In [7]:
print('Средняя ректальная температура:', selected_df.rectal_temperature.mean())
print('Медиана ректальной температуры:', selected_df.rectal_temperature.median())
print('Мода ректальной температуры:', selected_df.rectal_temperature.mode()[0])
print('Стандартное отклонение ректальной температуры:', selected_df.rectal_temperature.std())
print('Дисперсия ректальной температуры:', selected_df.rectal_temperature.var())
print('Первый квартиль ректальной температуры:', selected_df.rectal_temperature.quantile(0.25))
print('Третий квартиль ректальной температуры:', selected_df.rectal_temperature.quantile(0.75))

Средняя ректальная температура: 38.16791666666669
Медиана ректальной температуры: 38.2
Мода ректальной температуры: 38.0
Стандартное отклонение ректальной температуры: 0.7322886641121578
Дисперсия ректальной температуры: 0.5362466875871686
Первый квартиль ректальной температуры: 37.8
Третий квартиль ректальной температуры: 38.5


Температурные средние очень плотны друг к другу, а отклонения незначительны. Так и должно быть: в более широких температурных диапазонах теплокровные живые существа не выживают.

In [8]:
print('Средний пульс:', selected_df.pulse.mean())
print('Медиана пульса:', selected_df.pulse.median())
print('Мода пульса:', selected_df.pulse.mode()[0])
print('Стандартное отклонение пульса:', selected_df.pulse.std())
print('Дисперсия пульса:', selected_df.pulse.var())
print('Первый квартиль пульса:', selected_df.pulse.quantile(0.25))
print('Третий квартиль пульса:', selected_df.pulse.quantile(0.75))

Средний пульс: 71.91304347826087
Медиана пульса: 64.0
Мода пульса: 48.0
Стандартное отклонение пульса: 28.630556660735003
Дисперсия пульса: 819.7087747035575
Первый квартиль пульса: 48.0
Третий квартиль пульса: 88.0


Т.к. медиана пульса меньше его среднего арифметического, большинство лошадей имеет пульс меньше среднего. Значит среднее выростает за счет некоторого количества пиковых значений, которые могут претендовать на статус выбросов. Кроме этого, т.к. данные имеют большой разброс, а мода сильно удалена от других средних, можно предположить, что здесь наблюдается ситуация, когда есть много различных значений, отдельные из которых могут повторяться, но относительно небольшое количество раз. При таком распределении значений мода не очень показательна, т.к. ее значение  получилось в какой-то мере случайно (модой легко могло быть значение в совершенно другой части совокупности). Проверим эту гипотезу:

In [9]:
# Гипотеза подтверждается.
selected_df.pulse.value_counts().head()

48.0    28
60.0    25
40.0    18
88.0    12
52.0    12
Name: pulse, dtype: int64

In [10]:
print('Средний гематокрит:', selected_df.packed_cell_volume.mean())
print('Медиана гематокрита:', selected_df.packed_cell_volume.median())
print('Мода гематокрита:', selected_df.packed_cell_volume.mode()[0])
print('Стандартное отклонение гематокрита:', selected_df.packed_cell_volume.std())
print('Дисперсия гематокрита:', selected_df.packed_cell_volume.var())
print('Первый квартиль гематокрита:', selected_df.packed_cell_volume.quantile(0.25))
print('Третий квартиль гематокрита:', selected_df.packed_cell_volume.quantile(0.75))

Средний гематокрит: 46.29520295202952
Медиана гематокрита: 45.0
Мода гематокрита: 37.0
Стандартное отклонение гематокрита: 10.419334549704114
Дисперсия гематокрита: 108.56253245865784
Первый квартиль гематокрита: 38.0
Третий квартиль гематокрита: 52.0


Значения гематокрита распределены гораздо более "кучно", чем показатели пульса, т.к. стандартное отклонение существенно меньше. При этом мода и медиана близки друг к другу, значит в представленных данных нет односторонне проявленных крайних пиковых значений. Хотя бОльшее количество значений все же чуть меньше среднего. Мода здесь тоже не кажется показательным значением.

In [11]:
# Проверяем, какие значения встречаются наиболее часто и видим, что на самом деле у нас 2 моды, достаточно далеко
# отстоящие друг от друга (больше, чем на размер стандартного отклонения). И заодно еще раз убеждаемся
# в неплохой "кучности" результатов по этому показателю:
selected_df.packed_cell_volume.value_counts().head()

50.0    16
37.0    16
45.0    14
43.0    14
44.0    13
Name: packed_cell_volume, dtype: int64

In [12]:
print('Средний общий белок:', selected_df.total_protein.mean())
print('Медиана общего белка:', selected_df.total_protein.median())
print('Мода общего белка:', selected_df.total_protein.mode()[0])
print('Стандартное отклонение общего белка:', selected_df.total_protein.std())
print('Дисперсия общего белка:', selected_df.total_protein.var())
print('Первый квартиль общего белка:', selected_df.total_protein.quantile(0.25))
print('Третий квартиль общего белка:', selected_df.total_protein.quantile(0.75))

Средний общий белок: 24.456928838951317
Медиана общего белка: 7.5
Мода общего белка: 6.5
Стандартное отклонение общего белка: 27.475009470785064
Дисперсия общего белка: 754.8761454197289
Первый квартиль общего белка: 6.5
Третий квартиль общего белка: 57.0


Видим, что большинство значений общего белка тяготеет к нижней границе, ведь медиана равна 7.5, а первый квартиль - 6.5. При этом разброс значений очень большой. И среднее значение здесь вообще не показательно (так же, как непоказательна средняя зарплата по стране).

---

#### Задание 2. Работа с выбросами
<div class="alert alert-block alert-success">

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

</div>

---

Прежде всего, проверим, есть ли выбросы для **категориальных величин**:

In [13]:
# Видим, что для возраста вместо допустимых значений 1 и 2 есть значения 1 и 9.
selected_df.age.value_counts()

1    276
9     24
Name: age, dtype: int64

In [14]:
# Логично предположить, что вследствии ошибки ввода данных (вероятно, из-за случайной автозамены) 2 было
# заменено на 9 во всех случаях. Выполним обратную замену:
selected_df.age.replace(9, 2, inplace=True)

In [15]:
# Для параметра 'abdominal_distension' выбросов нет:
selected_df.abdominal_distension.value_counts()

1.0    76
3.0    65
2.0    65
4.0    38
Name: abdominal_distension, dtype: int64

In [16]:
# Для параметра 'surgery' выбросов нет:
selected_df.surgery.value_counts()

1.0    180
2.0    119
Name: surgery, dtype: int64

In [17]:
# Для параметра 'outcome' выбросов нет:
selected_df.outcome.value_counts()

1.0    178
2.0     77
3.0     44
Name: outcome, dtype: int64

Заодно видим, что лошадиная больничка, видимо, не сильно хорошая, ведь **выживает 59,53% "пациентов"** - (178/299)*100%.

Теперь посмотрим на возможные выбросы в **количественных данных**. Начнем с **ректальной температуры**:

In [18]:
# Т.к. нормальная ректальная температура лошади 37.8, похоже, что все значения в пределах допустимых.
selected_df.rectal_temperature.value_counts()

38.0    26
38.5    19
38.3    18
37.8    17
38.2    16
38.6    12
38.1    12
37.5    12
38.4    11
37.6     7
37.7     7
37.2     7
38.7     7
37.9     7
38.8     6
37.3     6
39.2     5
39.3     4
38.9     4
39.5     4
39.0     4
37.4     3
39.4     3
37.1     3
40.3     2
36.5     2
37.0     2
39.1     2
36.0     1
40.0     1
36.4     1
39.9     1
35.4     1
36.9     1
39.6     1
36.1     1
36.8     1
40.8     1
39.7     1
36.6     1
Name: rectal_temperature, dtype: int64

In [ ]:
# Т.к. температура - величина непрерывная, в случае бОльшего количества значений в нашем датафрейме
# их уже было бы нельзя оценить визуально. В таком случае мы бы выбрали диапазон допустимых значений
# (например, от 35 до 41) и отфильтровали бы наш датафрейм следующим образом:

selected_df[(selected_df.rectal_temperature < 35) | (selected_df.rectal_temperature > 41)]

# И уже потом решали бы, что делать со значениями, которые не попадают в выбранный диапазон.

Теперь посмотрим на **пульс**, помня, что нормальное значение 30-40 ударов в минуту для нетренированной лошади и 20-25 - для тренированной:

In [19]:
print('Всего уникальных значений пульса:', selected_df.pulse.nunique())
selected_df.pulse.value_counts()

# Количество значений пока позволяет провести визуальную оценку "на глаз".
# Что-то пульс какой-то повышенный у всех. Может лошади просто больны, а может это выбросы... Нужно изучить детальней.

Всего уникальных значений пульса: 52


48.0     28
60.0     25
40.0     18
88.0     12
52.0     12
100.0    11
44.0     11
72.0     11
120.0    10
42.0     10
80.0      8
64.0      8
96.0      8
84.0      8
54.0      7
56.0      7
50.0      6
66.0      6
68.0      5
104.0     5
70.0      4
76.0      4
92.0      4
78.0      3
90.0      3
108.0     3
86.0      3
36.0      3
112.0     3
130.0     2
140.0     2
124.0     2
114.0     2
45.0      2
150.0     2
30.0      2
146.0     1
136.0     1
184.0     1
164.0     1
129.0     1
65.0      1
132.0     1
49.0      1
98.0      1
38.0      1
160.0     1
128.0     1
110.0     1
46.0      1
75.0      1
82.0      1
Name: pulse, dtype: int64

Узнаем на [специализированом сайте](https://ru.qaz.wiki/wiki/Circulatory_system_of_the_horse), что: "Сердечно-сосудистая система лошади в высшей степени совместима с диапазоном частоты сердечных сокращений от 20 до 240 ударов в минуту". Исходя из этого считаем, что выбросы в данных по пульсу отсутствуют.

In [20]:
# При этом замечаем, что среди лошадей с пульсом выше 130 выжило всего 3, хотя количество таких лошадей - 10.
# Т.е. выживаемость у лошадей с высоким пульсом - сильно ниже средней (которая, как мы помним, 59,53%)

print('Всего лошадей с пульсом выше 130:', len(selected_df[selected_df.pulse > 130]))
print('Из них выжило:', len(selected_df[(selected_df.pulse > 130) & (selected_df.outcome == 1)]), 'А именно следующие:')
selected_df[(selected_df.pulse > 130) & (selected_df.outcome == 1)]

Всего лошадей с пульсом выше 130: 10
Из них выжило: 3 А именно следующие:


age  rectal_temperature  pulse  packed_cell_volume  total_protein  \
41     2                39.0  150.0                47.0            8.5   
103    2                38.0  140.0                39.0            5.3   
148    1                38.3  132.0                57.0            8.0   

     abdominal_distension  surgery  outcome  
41                    NaN      2.0      1.0  
103                   2.0      1.0      1.0  
148                   2.0      1.0      1.0

Т.е. очень высокий пульс лошадей, когда они находятся в состоянии относительного покоя (в больнице сильно не побегаешь), не способствует их выживанию.

Теперь посмотрим на **гематокрит**, помня, что его нормальный диапазон для здоровой лошади составляет от 30 до 50. Уровень повышается, когда нарушается кровообращение или когда животное становится обезвоженным.

In [21]:
# Видим, что тоже нет никаких заоблачных значений. И считаем все показатели допустимыми, помня, что лошади больны.
selected_df.packed_cell_volume.value_counts()

50.0    16
37.0    16
45.0    14
43.0    14
44.0    13
40.0    12
36.0    11
47.0    10
35.0    10
46.0     9
39.0     9
38.0     9
48.0     9
33.0     8
42.0     8
60.0     8
41.0     8
49.0     7
54.0     6
55.0     6
52.0     6
57.0     6
34.0     5
53.0     4
65.0     4
68.0     4
64.0     4
66.0     3
63.0     2
32.0     2
56.0     2
73.0     2
58.0     2
69.0     2
51.0     2
59.0     2
31.0     2
75.0     2
72.0     1
70.0     1
31.5     1
71.0     1
74.0     1
37.5     1
23.0     1
30.0     1
67.0     1
24.0     1
26.0     1
28.0     1
Name: packed_cell_volume, dtype: int64

In [22]:
# Убеждаемся в корректности нашего допущения о "нормальности" наличия в данных слишком высокого гематокрита,
# ведь лошади со слишком высоким гематокритом, как правило, не выживают. Для значения > 70 выжила 1 лошадь из 7.

selected_df[selected_df.packed_cell_volume > 70]

age  rectal_temperature  pulse  packed_cell_volume  total_protein  \
4      1                37.3  104.0                74.0            7.4   
35     1                 NaN  104.0                73.0            8.4   
49     1                37.2   84.0                73.0            5.5   
144    1                37.1   84.0                75.0           81.0   
151    1                39.3    NaN                75.0            NaN   
217    1                38.0   76.0                71.0           11.0   
231    1                39.5   92.0                72.0            6.4   

     abdominal_distension  surgery  outcome  
4                     NaN      2.0      2.0  
35                    3.0      2.0      3.0  
49                    2.0      2.0      2.0  
144                   4.0      1.0      3.0  
151                   2.0      1.0      2.0  
217                   NaN      1.0      1.0  
231                   1.0      2.0      2.0

И, наконец, посмотрим на **общий белок**, нормальный диапазон которого составляет от 30 до 50 (г/дл). Уровень повышается по мере нарушения кровообращения или обезвоживания животного.

In [23]:
# Здесь значений много, разброс между ними большой. Поэтому не будем пробовать оценить их "на глаз".

print('Всего уникальных значений общего белка:', selected_df.total_protein.nunique())
print('Минимальное значение общего белка:', selected_df.total_protein.min())
print('Максимальное значение общего белка:', selected_df.total_protein.max())

Всего уникальных значений общего белка: 81
Минимальное значение общего белка: 3.3
Максимальное значение общего белка: 89.0


Попытаемся понять, нормален ли такой разброс значений.

In [24]:
# Видим, что при совсем низких показателях общего белка (< 5 (г/дл)) лошади в шести случаях из девяти не выжили.
# Т.е. выживаемость - 33,3%, что существенно ниже среднего значения 59,53%.
# Поэтому также допускаем "нормальность" наличия таких низких показателей белка в данных.

selected_df[selected_df.total_protein <= 5]

age  rectal_temperature  pulse  packed_cell_volume  total_protein  \
74     2                 NaN    NaN                37.0            4.9   
84     1                37.8   60.0                40.0            4.5   
135    2                38.1  136.0                33.0            4.9   
142    1                 NaN  120.0                57.0            4.5   
191    2                 NaN  100.0                37.0            4.7   
210    1                37.9   68.0                45.0            4.0   
247    1                38.1   88.0                41.0            4.6   
255    2                38.8  184.0                33.0            3.3   
287    2                39.5   84.0                28.0            5.0   

     abdominal_distension  surgery  outcome  
74                    NaN      1.0      2.0  
84                    3.0      1.0      1.0  
135                   3.0      1.0      2.0  
142                   4.0      1.0      2.0  
191                   1.0      2.0      1.0  
210                   2.0      1.0      2.0  
247                   3.0      1.0      2.0  
255                   3.0      1.0      2.0  
287                   NaN      2.0      1.0

In [25]:
# Мы также видим, что лошадей с показателями общего белка от 5 до 10 (г/дл) - вообще бОльшая часть нашей выборки.
# И мы уже о чем-то подобном догадывались, ранее рассчитав медиану.
# Поэтому мы никак не можем считать эти значения выбросами. Просто лошадки болеют.

len(selected_df.total_protein[(selected_df.total_protein > 5) & (selected_df.total_protein < 10)])

176

In [26]:
# В связи с сильным повышением уровня общего белка также наблюдаем снижение уровня выживаемости до тех же 33,3%
# (выжили 2 из 6 лошадей). Поэтому также допускаем "нормальность" наличия этих показателей в данных.

selected_df[selected_df.total_protein > 80]

age  rectal_temperature  pulse  packed_cell_volume  total_protein  \
1      1                39.2   88.0                50.0           85.0   
14     1                38.2   76.0                46.0           81.0   
63     1                38.2  130.0                65.0           82.0   
90     1                38.0   52.0                53.0           86.0   
144    1                37.1   84.0                75.0           81.0   
208    1                37.8   88.0                64.0           89.0   

     abdominal_distension  surgery  outcome  
1                     2.0      1.0      3.0  
14                    1.0      1.0      1.0  
63                    4.0      2.0      3.0  
90                    NaN      2.0      1.0  
144                   4.0      1.0      3.0  
208                   3.0      1.0      3.0

In [27]:
# А вот умеренное повышение общего белка, оказывается, наоборот, способствует выживаемости.
# У лошадей со значениям общего белка от 70 до 80 (г/дл) выживаемость составляет 75%, что существенно выше среднего:

len_70_80 = len(selected_df[(selected_df.total_protein > 70) & (selected_df.total_protein < 80)])
len_70_80_survived = len(selected_df[(selected_df.total_protein > 70) & (selected_df.total_protein < 80) & (selected_df.outcome == 1)])

survived_percentage_70_80 = (len_70_80_survived / len_70_80) * 100
survived_percentage_70_80

75.0

Итого делаем вывод, что и по общему белку выбросов нет.

Какие-то у нас выбранные показатели очень правильные получились, совсем без выбросов. Если бы в домашнем задании было нужно обязательно показать навык удаления выбросов из датафрейма, это можно было бы сделать на примере 'rectal_temperature' с помощью следующего кода:

In [28]:
q1 = selected_df.rectal_temperature.quantile(0.25)
q3 = selected_df.rectal_temperature.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = selected_df[selected_df.rectal_temperature.between(lower_bound, upper_bound, inclusive=True)]
remove_outliers

age  rectal_temperature  pulse  packed_cell_volume  total_protein  \
0      1                38.5   66.0                45.0            8.4   
1      1                39.2   88.0                50.0           85.0   
2      1                38.3   40.0                33.0            6.7   
3      2                39.1  164.0                48.0            7.2   
4      1                37.3  104.0                74.0            7.4   
..   ...                 ...    ...                 ...            ...   
292    1                37.0   66.0                35.0            6.9   
294    1                38.5   40.0                37.0           67.0   
296    1                37.2   72.0                44.0            NaN   
297    1                37.5   72.0                60.0            6.8   
299    1                37.2   40.0                36.0           62.0   

     abdominal_distension  surgery  outcome  
0                     4.0      2.0      2.0  
1                     2.0      1.0      3.0  
2                     1.0      2.0      1.0  
3                     4.0      1.0      2.0  
4                     NaN      2.0      2.0  
..                    ...      ...      ...  
292                   3.0      1.0      2.0  
294                   1.0      2.0      1.0  
296                   3.0      2.0      3.0  
297                   3.0      1.0      2.0  
299                   NaN      1.0      3.0  

[226 rows x 8 columns]

Итого получился датафрейм с 226 строками, т.е. мы беспричинно удалили целых 74 строки, включая все строки, где 'rectal_temperature' == None. А эти строки жалко. Поэтому оставим их в датафрейме 'selected_df'.

---

#### Задание 3. Работа с пропусками
<div class="alert alert-block alert-success">

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

</div>

---

In [29]:
selected_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   age                   300 non-null    int64  
 1   rectal_temperature    240 non-null    float64
 2   pulse                 276 non-null    float64
 3   packed_cell_volume    271 non-null    float64
 4   total_protein         267 non-null    float64
 5   abdominal_distension  244 non-null    float64
 6   surgery               299 non-null    float64
 7   outcome               299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [30]:
# Рассчитаем количество и процент пропусков для каждого столбца
for col in selected_df.columns:
    nones_count = selected_df[col].isna().sum()
    nones_share = selected_df[col].isna().mean()
    print(f'{col} - {nones_count} пропусков, или {nones_share :.2%} от общего количества данных в столбце.')

age - 0 пропусков, или 0.00% от общего количества данных в столбце.
rectal_temperature - 60 пропусков, или 20.00% от общего количества данных в столбце.
pulse - 24 пропусков, или 8.00% от общего количества данных в столбце.
packed_cell_volume - 29 пропусков, или 9.67% от общего количества данных в столбце.
total_protein - 33 пропусков, или 11.00% от общего количества данных в столбце.
abdominal_distension - 56 пропусков, или 18.67% от общего количества данных в столбце.
surgery - 1 пропусков, или 0.33% от общего количества данных в столбце.
outcome - 1 пропусков, или 0.33% от общего количества данных в столбце.


In [31]:
selected_df[selected_df.outcome.isna()]

age  rectal_temperature  pulse  packed_cell_volume  total_protein  \
132    1                38.0   48.0                43.0           73.0   

     abdominal_distension  surgery  outcome  
132                   1.0      NaN      NaN

Принимая решение о замене пропусков, помним, что для категориальных данных часто лучшее решение - замена пропусков модой. Проводим такую замену для 'surgery' и 'outcome'.

In [32]:
selected_df.surgery.fillna(selected_df.surgery.mode()[0], inplace=True)
selected_df.outcome.fillna(selected_df.outcome.mode()[0], inplace=True)

У нас остаются 5 колонок с пропущенными значениями. Посмотрим, есть ли строки, у которых все значения в этих колонках пропущены:

In [33]:
selected_df.dropna(thresh=4).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   age                   296 non-null    int64  
 1   rectal_temperature    240 non-null    float64
 2   pulse                 276 non-null    float64
 3   packed_cell_volume    271 non-null    float64
 4   total_protein         267 non-null    float64
 5   abdominal_distension  244 non-null    float64
 6   surgery               296 non-null    float64
 7   outcome               296 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 20.8 KB


Т.к. целью заполнения пропусков является сохранение реальных данных, а не размножение фейковых данных, примем решение эти строки удалить, поскольку там для нас нет никакой полезной информации.

In [34]:
selected_df.dropna(thresh=4, inplace=True)

In [35]:
selected_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   age                   296 non-null    int64  
 1   rectal_temperature    240 non-null    float64
 2   pulse                 276 non-null    float64
 3   packed_cell_volume    271 non-null    float64
 4   total_protein         267 non-null    float64
 5   abdominal_distension  244 non-null    float64
 6   surgery               296 non-null    float64
 7   outcome               296 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 20.8 KB


Теперь подумаем, как заполнить пропуски в столбце 'abdominal_distension'. Для начала посмотрим, как распределились непустые значения помня, что пустых осталось 52 (296-244):

In [36]:
selected_df.abdominal_distension.value_counts()

1.0    76
3.0    65
2.0    65
4.0    38
Name: abdominal_distension, dtype: int64

Хотя abdominal_distension (вздутие живота) - значение категориальное, провести замену пропусков модой здесь мы не можем, т.к. пропусков много, и такая замена, увы, приведет к существенному искажению данных. Но мы помним из описания данных, что у лошадей со вздутыми животами, вероятно, будет наблюдаться повышенная болезненность. И у них также снижена перистальтика кишечника. Кроме этого, лошадям с сильным вздутием живота, вероятно, потребуется операция только для устранения вздутия. Исходя из этих знаний можно было бы заполнить пропуски в показателях вздутия на основании данных других колонок. Но из перечисленных колонок мы включили в наш датафрейм только данные по хирургии. Поэтому будем отталкиваться от них, выполнив замену пропусков модой 'abdominal_distension', для каждого из значений 'surgery'. Это тоже приведет к искажению данных, но в меньшей степени.

In [37]:
# Узнаем, какова же мода у 'abdominal_distension' при каждом из значений 'surgery':

print(selected_df[selected_df.surgery == 1].abdominal_distension.mode()[0])
print(selected_df[selected_df.surgery == 2].abdominal_distension.mode()[0])

3.0
1.0


In [38]:
# Напишем функцию для замены

def replace_with_mode(row):
    '''Функция заменяет пустые значения в колонке 'abdominal_distension', в зависимости
    от значений  в колонке'surgery' '''
    
    if row['surgery'] == 1 and pd.isna(row['abdominal_distension']):
        return 3
    elif row['surgery'] == 2 and pd.isna(row['abdominal_distension']):
        return 1
    else:
        return row['abdominal_distension']

In [39]:
# Применим функцию к датафрейму

selected_df['abdominal_distension'] = selected_df.apply(replace_with_mode, axis = 1)

In [40]:
selected_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   age                   296 non-null    int64  
 1   rectal_temperature    240 non-null    float64
 2   pulse                 276 non-null    float64
 3   packed_cell_volume    271 non-null    float64
 4   total_protein         267 non-null    float64
 5   abdominal_distension  296 non-null    float64
 6   surgery               296 non-null    float64
 7   outcome               296 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 20.8 KB


Теперь поработаем с пропусками для количественных величин: **'rectal_temperature'**, **'pulse'**, **'packed_cell_volume'**, **'total_protein'**. Т.к. нам неизвестны какие-то зависимости между каждым из этих показателей и другими, представленными в датафрейме, ничего не остается, как заменить пропуски в каждой из этих колонок соответствующей медианой:

In [41]:
selected_df.rectal_temperature.fillna(selected_df.rectal_temperature.median(), inplace=True)
selected_df.pulse.fillna(selected_df.pulse.median(), inplace=True)
selected_df.packed_cell_volume.fillna(selected_df.packed_cell_volume.median(), inplace=True)
selected_df.total_protein.fillna(selected_df.total_protein.median(), inplace=True)

Убеждаемся, что в датафрейме не осталось пропусков:

In [42]:
# Это можно сделать так:
selected_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   age                   296 non-null    int64  
 1   rectal_temperature    296 non-null    float64
 2   pulse                 296 non-null    float64
 3   packed_cell_volume    296 non-null    float64
 4   total_protein         296 non-null    float64
 5   abdominal_distension  296 non-null    float64
 6   surgery               296 non-null    float64
 7   outcome               296 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 20.8 KB


In [43]:
# Или так:
for col in selected_df.columns:
    nones_count = selected_df[col].isna().sum()
    print(f'{col} - {nones_count} пропусков.')

age - 0 пропусков.
rectal_temperature - 0 пропусков.
pulse - 0 пропусков.
packed_cell_volume - 0 пропусков.
total_protein - 0 пропусков.
abdominal_distension - 0 пропусков.
surgery - 0 пропусков.
outcome - 0 пропусков.


---

**Дополнительная часть (необязательная):** Выполнить задания 1-3 для всего набора данных.

Пожалуй, в этот раз воздержусь от выполнения дополнительной части. И так много получилось :)
А впереди - много другого интересного материала и мало времени до дедлайна.